### PR TITLE
Add production server

### PR DESCRIPTION
**Pull request by chrisdevereux**
* Enhances build command to produce minified/crushed/optimized js/asset/css files
* Uses built asset bundles when running in production mode rather than the devserver (fixes flash of unstyled content)
* Uses `<link>` rather than `<style>` for css in production
* Gzips html and js returned from server

Yields a 10x reduction in js bundle size when running in production

### CODE REVIEW

File: src/cli/local-commands/build.command.ts [hunk 1]
```diff
@@ -19,9 +22,33 @@ export async function handler() {
...
+      pages: getEntrypointFiles(),
+      plugins,
+      devServer: false
+    }))
+
+    await runCompiler(compiler)
   })
 
+  function getEntrypointFiles() {
+    const filePattern = path.join('src', 'pages', '**', '*.@(t|j)s?(x)')
```
Comment (robotdevil) [inline src/cli/local-commands/build.command.ts]: Didn't you have to take this path.join out to solve the issue with Windows path - is this re-instating it?

File: src/lib/bundler/getWebpackConfig.test.ts [hunk 2]
```diff
@@ -1,7 +1,11 @@
 import { getWebpackConfig } from './getWebpackConfig'
 
 describe('getWebpackConfig()', () => {
-  it('should generate webpack config', () => {
-    getWebpackConfig({ pages: ['foo.js'], plugins: [] })
+  it('should generate devserver webpack config', () => {
+    getWebpackConfig({ pages: ['foo.js'], plugins: [], devServer: true })
```
Comment (robotdevil) [inline src/lib/bundler/getWebpackConfig.test.ts:5]: Sorry - might be being thick - do these tests actually check the resulting webpack config to confirm it is the expected one?

File: src/lib/server/startProductionServer.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import * as express from 'express'
+import * as path from 'path'
+import * as compression from 'compression'
+import { pick } from 'lodash'
+import { renderHtmlWrapper } from './renderHtmlWrapper'
+import { Config } from './Config'
+import { enforceCustomerFacingHttps } from './enforceHttps'
+
+export interface DevServerOpts {
```
Comment (robotdevil) [inline src/lib/server/startProductionServer.ts]: ProductionServerOpts?

File: src/lib/bundler/getWebpackConfig.test.ts (hunk 2 above)
Comment (chrisdevereux) [reply src/lib/bundler/getWebpackConfig.test.ts:5]: They don't. It basically just confirms that it doesn't throw.Really testing that the config is correct would probably require end-end tests (eg. selenium tests against a running server) which we don't currently have but probably should.

File: src/lib/server/startProductionServer.ts (hunk 3 above)
Comment (chrisdevereux) [reply src/lib/server/startProductionServer.ts]: 👍

File: src/cli/local-commands/build.command.ts (hunk 1 above)
Comment (chrisdevereux) [reply src/cli/local-commands/build.command.ts]: 👍 There's also some duplication between the build and run commands which should probably be extracted

**Review by robotdevil — DISMISSED** [review]

**Review by robotdevil — APPROVED** [review]

**Pull request merged**